add mixed data type support for GroupNorm

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #81852

1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16,
the input/output and gamma/beta will all be in bfloat16.

cc @VitalyFedyunin @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

mingfeima committed:
add mixed data type support for GroupNorm

1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16,
the input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for GroupNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16,
the input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

frank-wei commented on aten/src/ATen/native/cpu/group_norm_kernel.cpp:
```diff
@@ -166,7 +165,7 @@ void GroupNormKernelImplChannelsLastInternal(
     int64_t HxW,
     int64_t group,
-    T eps,
+    double eps,
```
why eps needs double here? What about float?

mingfeima replied:
revert this to align with the cuda kernels https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/group_norm_kernel.cu#L554

frank-wei commented on aten/src/ATen/native/cpu/group_norm_kernel.cpp:
```diff
@@ -348,8 +347,8 @@ void GroupNormKernelImplChannelsLastInternal(
           const int64_t c = g * D + d;
-          scale_ptr[c] = rstd_val * (gamma_null ? T(1) : gamma_data[c]);
-          bias_ptr[c] = -scale_ptr[c] * mean_val + (beta_null ? T(0) : beta_data[c]);
+          scale_ptr[c] = rstd_val * (gamma_null ? PT(1) : gamma_data[c]);
```
scale_ptr is type T, why PT(1) is used here?

mingfeima replied:
`scale_ptr` is of type `T` but `gamma_data` and `beta_data` is of type `PT`.

when input is bfloat16, the params type for mixed data type will be:

* input is bfloat16, params (gamma/beta) is bfloat16: `T` and `PT` are both `BFloat16`.
* input is bfloat16, params (gamma/beta) is float: `T` is `BFloat16` and `PT` is `float`. And compiler would require that in the phrase of `gamma_null ? PT(1) : gamma_data[c]`, `gamma_data[c]` and 1 should have the same type.

mingfeima committed:
Update on "add mixed data type support for GroupNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16,
the input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for GroupNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16,
the input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for GroupNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16,
the input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for GroupNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16,
the input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for GroupNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16,
the input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for GroupNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16,
the input/output and gamma/beta will all be in bfloat16.

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for GroupNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16,
the input/output and gamma/beta will all be in bfloat16.

cc @VitalyFedyunin jgong5 @XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for GroupNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16,
the input/output and gamma/beta will all be in bfloat16.

cc @VitalyFedyunin jgong5 @XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima committed:
Update on "add mixed data type support for GroupNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16,
the input/output and gamma/beta will all be in bfloat16.

cc @VitalyFedyunin jgong5 @XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented:
@pytorchbot merge

malfet commented on aten/src/ATen/native/cpu/group_norm_kernel.cpp:
```diff
@@ -393,19 +392,30 @@ void GroupNormKernelImpl(
       AT_DISPATCH_FLOATING_TYPES_AND(ScalarType::BFloat16, X.scalar_type(), "GroupNormKernelImpl", [&]() {
-        GroupNormKernelImplInternal<scalar_t>(
-            X, gamma, beta, N, C, HxW, group, static_cast<scalar_t>(eps), Y, mean, rstd);
+        if (mixed_type) {
+          GroupNormKernelImplInternal<BFloat16, float>(
+              X, gamma, beta, N, C, HxW, group, static_cast<scalar_t>(eps), Y, mean, rstd);
+        } else {
```
Hmm, this will add this if statement for all datatypes, but it makes sense only for float16, isn't it? Can we use a constexpr or something to restrict this one only for BFloat16 scalar type?

jgong5 replied:
Yes, users can create mixed dtype cases for other data types than bfloat16+float, even though these are not typical. Maybe a simpler way to fix it is to move the dtype check of bfloat16 and mixed type check out and before `AT_DISPATCH_FLOATING_TYPES_AND`.

mingfeima replied:
changed the mixed_dtype pattern into the following:
```C++
        using param_t = at::opmath_type<scalar_t>;
        if (mixed_type) {
          GroupNormKernelImplInternal<scalar_t, param_t>(
              X, gamma, beta, N, C, HxW, group, static_cast<scalar_t>(eps), Y, mean, rstd);
        } else {
          GroupNormKernelImplInternal<scalar_t, scalar_t>(
              X, gamma, beta, N, C, HxW, group, static_cast<scalar_t>(eps), Y, mean, rstd);
        }
```
We will enable `mixed_dtype` with float16 in near future, so `mixed_dtype` will be like:
* bfloat16 + float32
* float16 + float32

And for float16, we won't have to do anything special, just replacing the `AT_DISPATCH_FLOATING_TYPES_AND` will do the entire job.

malfet reviewed (approved):
Small nit on the perf, we introduce unnecessary `if (mixedtype)` for all dtypes but BFloat16, perhaps one can refactor this code as:
```
template<scalar_t>
void CallGroupNorm(...)
```
And specialize it for BFloat16 to call `mixed_type`

mingfeima committed:
Update on "add mixed data type support for GroupNorm"


1. If user uses amp to run bfloat16 models, `torch.autocast` will
keep module paramters in acc dtype which will leave `gamma` and`beta`
in float while input/output will be in bfloat16.

2. If user explicitly cast the model to bfloat16,
the input/output and gamma/beta will all be in bfloat16.

cc @VitalyFedyunin jgong5 @XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented:
> Small nit on the perf, we introduce unnecessary `if (mixedtype)` for all dtypes but BFloat16, perhaps one can refactor this code as:
> 
> ```
> template<scalar_t>
> void CallGroupNorm(...)
> ```
> 
> And specialize it for BFloat16 to call `mixed_type`

We will add minxed_type for float16 in near future for the coming new platform of Xeon (Granite Rapids).

Also I noticed that `vec_scalar_t` is a duplicate now, will have it replaced with `opmath_t`.

mingfeima commented:
@pytorchbot merge
